File: solutions/clojure/sieve/2/src/sieve.clj
(ns sieve)

(defn multiple-of? [n m]
  (->> m
       (iterate #(- % n))
       (drop-while pos?)
       first
       zero?))

(def not-multiple-of? (complement multiple-of?))

(defn sieve-step
  [primes [fst & rst]]
  (if fst
    (recur (conj primes fst)
           (filter (partial not-multiple-of fst) rst))
    primes))

(defn sieve
  [n]
  (sieve-step [] (range 2 (inc n))))

File: solutions/scala/sieve/2/src/main/scala/Sieve.scala
object Sieve {
  def primes(limit: Int): List[Int] =
    if limit < 2 then List()
    else
      val isPrime = Array.fill(limit + 1)(true)
      isPrime(0) = false
      isPrime(1) = false

      var number = 2
      while number * number <= limit do
        if isPrime(number) then
          var multiple = number * number
          while multiple <= limit do
            isPrime(multiple) = false
            multiple += number

        number += 1

      (2 to limit).filter(isPrime).toList
}
